internal/databases/mysql/backup_push_handler.go

```diff
@@ -1,14 +1,89 @@
 package mysql
 
 import (
+	"bufio"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+
+	"github.com/spf13/viper"
 	"github.com/wal-g/tracelog"
 	"github.com/wal-g/wal-g/internal"
 	"github.com/wal-g/wal-g/utility"
-	"os/exec"
 )
 
-func HandleBackupPush(uploader *internal.Uploader, backupCmd *exec.Cmd) {
-	uploader.UploadingFolder = uploader.UploadingFolder.GetSubFolder(utility.BaseBackupPath)
+const (
+	XtrabackupExtraLsnDir    = "--extra-lsndir"
+	XtrabackupIncrementalLSN = "--incremental-lsn"
+	XtrabackupCheckpoints    = "xtrabackup_checkpoints"
+)
+
+func HandleBackupPush(uploader *internal.Uploader, backupCmd *exec.Cmd, isFullBackup bool) {
+	maxDeltas := viper.GetInt(internal.DeltaMaxStepsSetting)
+	var err error
+	var previousBackupName string
+	previousBackupSentinelDto := internal.BackupSentinelDto{}
+	incrementCount := 1
+
+	folder := uploader.UploadingFolder
+	uploader.UploadingFolder = folder.GetSubFolder(utility.BaseBackupPath)
+	if maxDeltas > 0 && !isFullBackup && isXtrabackupCmd(backupCmd) {
+		previousBackupName, err = internal.GetLatestBackupName(folder)
+		if err != nil {
+			if _, ok := err.(internal.NoBackupsFoundError); ok {
+				tracelog.InfoLogger.Println("Couldn't find previous backup. Doing full backup.")
+			} else {
+				tracelog.ErrorLogger.FatalError(err)
+			}
+		} else {
+			previousBackup := internal.NewBackup(uploader.UploadingFolder, previousBackupName)
+			previousBackupSentinelDto, err = previousBackup.GetSentinel()
+			tracelog.ErrorLogger.FatalOnError(err)
+
+			if previousBackupSentinelDto.IncrementCount != nil {
+				incrementCount = *previousBackupSentinelDto.IncrementCount + 1
+			}
+
+			if incrementCount > maxDeltas {
+				tracelog.InfoLogger.Println("Reached max delta steps. Doing full backup.")
+				previousBackupSentinelDto = internal.BackupSentinelDto{}
+				incrementCount = 1
+			} else if previousBackupSentinelDto.BackupFinishLSN == nil {
+				tracelog.InfoLogger.Println("LATEST backup was made without support for delta feature. Fallback to full backup with LSN marker for future deltas.")
+				previousBackupSentinelDto = internal.BackupSentinelDto{}
+				incrementCount = 1
+			} else {
+				tracelog.InfoLogger.Printf("Delta backup from LSN %v. \n", *previousBackupSentinelDto.BackupFinishLSN)
+			}
+		}
+	} else {
+		tracelog.InfoLogger.Println("Doing full backup.")
+	}
+
+	createAndPushBackup(uploader, backupCmd, previousBackupSentinelDto, incrementCount)
+}
+
+func createAndPushBackup(
+	uploader *internal.Uploader, backupCmd *exec.Cmd,
+	previousBackupSentinelDto internal.BackupSentinelDto,
+	incrementCount int,
+) {
+	extraLsnDir := path.Join("/tmp", "extra_lsn")
+	isXtrabackup := isXtrabackupCmd(backupCmd)
+	fromLSN, previousBackupName := previousBackupSentinelDto.BackupFinishLSN, previousBackupSentinelDto.IncrementFullName
```

Comment:
LSN in MySQL (innodb) has meaning only inside the single instance.
Backups may be taken from different instances (primary or one of secondary replicas).
So if current backup and previous are from different instances - we can rely on the LSN.

But we still can rely on LSN as long as we are taking backup from the same server